### DISCUSSION TITLE
Parameterising CasADi functions after generating C code

### DISCUSSION THREAD

**Discussion by bulatkhusainov** [casadi/casadi#3544] [Q&A] [answered]
I am using CasADi with acados and looking into different ways of parameterizing dynamics/objective/constraints after compiling the controller. The controller is implemented as a Simulink s-function.

So far I have been able to parameterize the controller by declaring parameters as a vector of MX variables and adding this vector to  acados_ocp_model (`ocp_model.set('sym_p', model.sym_p`). Time varying parameters are then fed as an input to the sfunction. This is a standard parametrization method presented in many acados examples. I understand from CasADi perspective, parameters are treated as additional inputs.  While this way of parametrizing is functional, it significantly slows down the MPC controller for my model, the solve time went up by the factor of 2.

The main issue is that I need to parametrize the objective with a very long vector. The vector is the same for all prediction nodes. However, acados treats parameterization of every prediction node separately, which increases the total number of parameters significantly and the number of MX variables grows. 

For my application the vector of parameters is known at the beginning of the simulation and not changing while the controller is running. So I was thinking of taking advantage of that and finding a way of parameterizing CasADi dynamics/objective/constraints functions directly, bypassing the acados method of parametrizing. 

The idea is to create a custom CasADi function which will have parameters of interest as a part of its evaluation, so that we can call this function for symbolic evaluations when specifying dynamics/objective/constraints. One implementation of this idea I considered was creating a custom function with `f = external('f', './gen.so');`. The intention was that the evaluation function within gen.so will access a global parameter variable, which will be manipulated from another part of application.

Does `external` function work for Windows? I tried creating an external function by linking a *.dll file, in a similar way to *.so files presented in CasADi documentation, but got the following error: 
> Assertion "handle_!=0" failed: CommonExternal: Cannot open "./gen.dll". Error code (WIN32): 193

Even if we can make  `external` function work for Windows, I am not sure if it will be supported for code generation. So I would appreciate any help/suggestions on:

- feasibility of the idea described above
- any other ideas on parametrizing the C code generated by CasADi post compilation.

**Comment by jgillis** [comment, marked as answer]
In general, this may be your answer:
```
/** \brief Construct matrix with a given sparsity and a file with nonzeros */
    MX(const Sparsity& sp, const std::string& fname);
```

In particular, about the external strategy.. External does work for Windows (with that exact syntax), and is supported for codegen.
Maybe some dependency of your dll is not found in the PATH? Try dependencywalker

**Reply by bulatkhusainov** [reply]
Thanks for getting back.

After you confirming that `f = external('f', './sharedlib.extension');` works for Windows, I investigated my *.dll further and realized that the library was compiled for 32 bit instead of 64 bit. Compiling for 64 bit resolved the issue and `external` function worked as expected. It is also great to see that C code generated by CasADi is architecture independent, as my code needs to run both on Windows and Linux, so there is a requirement to link both *.dll and *.so files.

Regarding `MX(const Sparsity& sp, const std::string& fname);` , my understanding is that this function loads data from files, although it is not entirely clear how it works, e.g. whether the file is read during compilation or runtime. Are there any examples on this? Could not find anything in casadi-example_pack or the main CasADi GitHub repo, may be I am missing something obvious here. But it was encouraging to see that the same functionality is available in MATLAB: `new_obj = MX(Sparsity sp, char fname)`.

**Reply by jgillis** [reply]
The file is read during runtime, when the code initializes. Just try and check the generated code. There is no example or documentation yet, feel free to suggest some.

**Reply by bulatkhusainov** [reply]
Will give it a try.

To summarise, coming back to the original problem of parametrizing CasADi functions post code generation, we end up with two possible directions:

- Load parameters from a file with `MX(const Sparsity& sp, const std::string& fname);`
- (i) Create a custom CasADi function, which will contain function parameters; (ii) compile the function as a shared library;  (iii) Link the shared library with` f = external('f', './sharedlib.extension');`

I am going to experiment with both options and chose the most suitable for my application.

Many thanks for the guidance.
